src/Npgsql/NpgsqlReadBuffer.Stream.cs

```diff
@@ -158,12 +158,50 @@ async Task<int> ReadLong(ValueTask<int> task, bool async)
                 return read;
             }
 
+#if !NET461 && !NETSTANDARD2_0
+            public override int Read(Span<byte> span)
+                => Read(span.ToArray(), CancellationToken.None, false).GetAwaiter().GetResult();
+
+            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+                => Read(buffer, cancellationToken, true);
+
+            ValueTask<int> Read(Memory<byte> buffer, CancellationToken cancellationToken, bool async)
+            {
+                CheckDisposed();
+
+                if (cancellationToken.IsCancellationRequested)
+                    return new ValueTask<int>(Task.FromCanceled<int>(cancellationToken));
+
+                var count = Math.Min(buffer.Length, _len - _read);
+
+                if (count == 0)
+                    return new ValueTask<int>(0);
+
+                var task = _buf.ReadBytes(buffer.Slice(0, count), async);
+                if (task.IsCompleted) // This may be a bug in the new version of ValueTask
```

Comment:
Let's remove this micro-optimization altogether and make this method async (including the byte[] overloads). This was done in a time where there was considerable overhead for async methods completing synchronously, so I did my best to avoid async methods in synchronously-returning paths. This specific perf has improved a lot, and this code path is really not hot enough to warrant anything like this.

Reply:
Removed.